src/libopensc/card-dtrust.c

```diff
@@ -438,6 +472,86 @@ dtrust_compute_signature(struct sc_card *card, const u8 *data,
 	return r;
 }
 
+static int
+_dtrust_compute_shared_value(struct sc_card *card,
+		const u8 *crgram, size_t crgram_len,
+		u8 *out, size_t outlen)
+{
+	int r;
+	struct sc_apdu apdu;
+	u8 *sbuf = NULL;
+
+	if (card == NULL || crgram == NULL || out == NULL) {
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+	LOG_FUNC_CALLED(card->ctx);
+	sc_log(card->ctx, "CardOS compute shared value: in-len %" SC_FORMAT_LEN_SIZE_T "u, out-len %" SC_FORMAT_LEN_SIZE_T "u", crgram_len, outlen);
```

Comment:
```suggestion
	sc_log(card->ctx, "D-Trust compute shared value: in-len %" SC_FORMAT_LEN_SIZE_T "u, out-len %" SC_FORMAT_LEN_SIZE_T "u", crgram_len, outlen);
```

Reply:
or if the function matches what is in the cardos, it might be better to move it to some shared file too.

Reply:
I don't have access to CardOS specification. Just the documentation of the D-Trust card specifies it should be a CardOS-specific command.

Do you have any suggestion where to place this function?

I doubt that the CardOS driver would be the right place as it introduces a dependency between two different card drivers. The CardOS driver is intended to be used for plain CardOS cards (at least as I understand the code). In contrast the D-Trust card is just a pre-initialized PKCS#15 card built on top of CardOS.

Reply:
Something like `card-caros-common.c` might be a good place. We use this semantics when sharing code between different drivers or their versions.

Reply:
> Something like `card-caros-common.c` might be a good place. We use this semantics when sharing code between different drivers or their versions.

:white_check_mark: thanks, fixed